scanpy/preprocessing/combat.py

```diff
@@ -0,0 +1,161 @@
+    adata : AnnData object 
+    key: str
+        key to a categorical annotation from adata.obs that will be used for batch effect removal
+    copy: bool
+        wether to update the adata object or to copy it
+
+    Returns
+    -------
+    Depending on the value of copy, either returns an updated AnnData object or modifies the passed one
+    """
+    
+    # only works on dense matrices so far
+    if issparse(adata.X): 
+        X = adata.X.A.T
+    else: 
+        X = adata.X.T
+    data = pd.DataFrame(data = X, index = adata.var_names,
+                        columns= adata.obs_names)
+    
+    # construct a pandas series of the batch annotation
+    batch = pd.Series(adata.obs[key])
+    model = pd.DataFrame({'batch': batch})
+    batch_items = model.groupby("batch").groups.items()
+    batch_levels = [k for k, v in batch_items]
+    batch_info = [v for k, v in batch_items]
+    n_batch = len(batch_info)
+    n_batches = np.array([len(v) for v in batch_info])
+    n_array = float(sum(n_batches))
+
+    # drop intercept
+    drop_cols = [cname for cname, inter in  ((model == 1).all()).iteritems() if inter == True]
+    model = model[[c for c in model.columns if not c in drop_cols]]
+    design = design_mat(model, batch_levels)
+
+    # standardize across genes using a pooled variance estimator
+    sys.stderr.write("Standardizing Data across genes.\n")
+    B_hat = np.dot(np.dot(la.inv(np.dot(design.T, design)), design.T), data.T)
+    grand_mean = np.dot((n_batches / n_array).T, B_hat[:n_batch,:])
+    var_pooled = np.dot(((data - np.dot(design, B_hat).T)**2), np.ones((int(n_array), 1)) / int(n_array))
+    if np.sum(var_pooled == 0) > 0:
+        print('Found {} genes with zero variance.'.\
+          format(np.sum(var_pooled == 0)))
+    stand_mean = np.dot(grand_mean.T.reshape((len(grand_mean), 1)), np.ones((1, int(n_array))))
+    tmp = np.array(design.copy())
+    tmp[:,:n_batch] = 0
+    stand_mean  += np.dot(tmp, B_hat).T
+
+    # need to be a bit careful with the zero variance genes
+    s_data = np.where(var_pooled == 0, 0, \
+        ((data - stand_mean) / np.dot(np.sqrt(var_pooled), np.ones((1, int(n_array))))))
+    s_data = pd.DataFrame(s_data, index=data.index, columns=data.columns)
+
+    # fitting the parameters
+    sys.stderr.write("Fitting L/S model and finding priors\n")
+    batch_design = design[design.columns[:n_batch]]
+    gamma_hat = np.dot(np.dot(la.inv(np.dot(batch_design.T, batch_design)), batch_design.T), s_data.T)
+    delta_hat = []
+
+    for i, batch_idxs in enumerate(batch_info):
+        delta_hat.append(s_data[batch_idxs].var(axis=1))
+
+    gamma_bar = gamma_hat.mean(axis=1) 
+    t2 = gamma_hat.var(axis=1)
+    a_prior = list(map(aprior, delta_hat))
+    b_prior = list(map(bprior, delta_hat))
+
+    sys.stderr.write("Finding parametric adjustments\n")
+    gamma_star, delta_star = [], []
+    for i, batch_idxs in enumerate(batch_info):
+        temp = it_sol(s_data[batch_idxs], gamma_hat[i],
```

Comment:
Mind your indentation. Generally, we'd indent along increments of 4 spaces and not along the position where a parenthesis opens. Here, you added a space that created misalignment.

Reply:
I don't get this comment - indentation between line 97 and 98 is 4 spaces, so I assume you mean the indentation between lines 98 and 99?

Reply:
yup, that one:

```py
temp = it_sol(
    s_data...,
    delta_hat...,
)
```